base/math.jl

```diff
@@ -911,13 +911,17 @@ function modf(x::Float64)
 end
 
 @inline function ^(x::Float64, y::Float64)
+    yint = unsafe_trunc(Int, y)
+    y == yint && return x^yint
```

Comment:
I don't think it is valid to call `unsafe_trunc` on values that might not be in range.

For example:

```
julia> f() =  Base.unsafe_trunc(Int, 10.0^307)
f (generic function with 1 method)

julia> @code_llvm f()
;  @ REPL[9]:1 within `f'
define i64 @julia_f_688() {
top:
  ret i64 undef
}
```

If you use the value after that, I think you run into undefined behavior.

Reply:
I think it is ok (and y == yint should fail for large values)

```
julia> @code_native Base.unsafe_trunc(Int, 10.0^307) # as opposed to f() which was constant-folded?
```

Reply:
Why do you think it is ok when e.g. the docstring for `unsafe_trunc` says:

>  If the value is not representable by T, an arbitrary value will be returned.

and what I showed above?

Reply:
the theory is that `unsafe_trunc` will return a correct result any time `y` can be exactly converted to an integer, and if `y` can't be exactly converted, no integer value can equal `y`. Thus even though I am relying on undefined behavior, it will produce the correct result.

Reply:
From what I can see `unsafe_trunc` (or `fpstoi` (https://llvm.org/docs/LangRef.html#fptosi-to-instruction)) returns a poison value (https://llvm.org/docs/LangRef.html#poisonvalues) whenever it is called with an argument out of range. And most operations on a poison value mean undefined behavior where at that point you cannot reason about your program anymore. But I am not an expert, I am just reading through the manuals.

Reply:
Why not compare `y == round(y)`?

Reply:
What about `isinteger(y)`?

Reply:
keno confirms your analysis. I was forgetting how viscous LLVM is allowed to be when poison values are involved.

Reply:
> What about `isinteger(y)`?

That gets `NaN` correct.

Reply:
I was relying on what LLVM actually does (on x86-64). round(y) is kind of bad as it produces large code and can throw an exception, unlike unsafe_trunc at least on x86-64. It's just one assembly instruction, always returning Int64 (even with Int, or Int32, thus presumably on 32-bit). Do we need to worry about (potential) LLVM undefined behavior, that I wasn't up-to-speed on, that may only apply to platforms Julia doesn't support (yet)?

Reply:
> I was relying on what LLVM actually does (on x86-64)

You looked at the generated code for one special case. You cannot infer "what LLVM actually does" based on that. You need to read the manual and see what is allowed etc.

Reply:
>Why do you think it is ok when e.g. the docstring for unsafe_trunc says:
>> If the value is not representable by T, an arbitrary value will be returned.

I'm ok with arbitrary, as long as it's Int64. I tried to dig into what actually happens, and it leads to INTRINSIC fptosi, that and this "poison value" business (while intriguing, I started looking into), is above my pay-grade (which is zero for this project).

```
julia> @code_llvm Base.unsafe_trunc(Int, 10.0^307)  # seems ok, unlike your f() function, not sure why you got "ret i64 undef"
;  @ float.jl:312 within `unsafe_trunc`
define i64 @julia_unsafe_trunc_792(double %0) #0 {
top:
  %1 = fptosi double %0 to i64
  %2 = freeze i64 %1
  ret i64 %2
}
```

Reply:
I am using Julia 1.6 and it seems things have changed a bit on a newer LLVM. The fact that there is a `freeze` there might actually help us...

Reply:
This really seems ok, since: "The ‘freeze’ instruction is used to stop propagation of undef and poison values. [..] If the argument is undef or poison, ‘freeze’ returns an arbitrary, but fixed, value of type ‘ty’."

Reply:
Ok after https://github.com/JuliaLang/julia/pull/38977 this is actually safe. Complaint withdrawn :)

Reply:
Never mind, further investigation shows that Julia puts special stuff in to make sure we don't have undefined behavior for stuff like this.

Reply:
It seems like this might depend on what you think you mean by "safe". The `y == yint` test is depending on an undefined value, so the result of that test is undefined also. The branch that it selects to take won't cause the program to abort with spooky action-at-a-distance, but it is not obligated to do anything useful.

Reply:
I checked and the "‘freeze’ returns an arbitrary, but fixed, value of type ‘ty’" but for out of range it always returns -9223372036854775808 (for ty == Int64). Still, it doesn't matter, it could return any Int64 value, even a random one, and the code would still work.

Reply:
Seconding the comments on avoiding [~nasal demons~ undefined behavior](http://catb.org/jargon/html/N/nasal-demons.html) because the compiler can do anything at that point. Moreover, compilers are allowed to assume that undefined behavior doesn't happen and e.g. optimize away things that would lead to undefined behavior instead of propagating the undefined result as poison.  If you're depending on Julia putting in special, guaranteed-to-propagate values, I would make sure that's actually documented somewhere and not itself undefined behavior....

Reply:
I thought the discussion established that there is no UB anymore. Also see the docs for `unsafe_trunc`

Reply:
Do we know that this test is actually safe to assume works here?